source/simulator/core.cc

```diff
@@ -2241,6 +2256,297 @@ namespace aspect
+                                                                                    std::max(0.9 * std::fabs(newton_residual*newton_residual)/(newton_residual_old*newton_residual_old),0.9*parameters.linear_stokes_solver_tolerance*parameters.linear_stokes_solver_tolerance));
+                            }
+                        }
+                      pcout << std::endl << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << ":" << std::fabs(newton_residual-stokes_residual)/newton_residual_old << "," << stokes_residual << "," << newton_residual << ", " << newton_residual_old << std::endl;
+
+                    }
+                }
+              build_stokes_preconditioner();
+
+
+              stokes_residual = solve_stokes();
+
+              /* The following block may be useful for when the newton solver fails. Is this somewthing we want?
+                            try
+                            {
+                                stokes_residual = solve_stokes();
+                            }
+                            catch (...)
+                            {
+                              // first get out of the timing section
+                              computing_timer.exit_section();
+                              // Newton solver failed. Try Picard instead.
+                              use_picard = true;
+                              parameters.newton_theta = 0;
+                                rebuild_stokes_matrix = rebuild_stokes_preconditioner = assemble_newton_stokes_matrix = true;
+                                assemble_stokes_system();
+                                parameters.linear_stokes_solver_tolerance = piccard_linear_stokes_solver_tolerance;
+                                pcout << std::endl << "   Solve failed, set Use Picard to true. and the linear solver tolerance is set to "
+                                  << parameters.linear_stokes_solver_tolerance << std::endl;
+                                stokes_residual = solve_stokes();
+                            }*/
+
+              newton_residual_velo = system_rhs.block(introspection.block_indices.velocities).l2_norm();
+              newton_residual_pres = system_rhs.block(introspection.block_indices.pressure).l2_norm();
+              newton_residual = std::sqrt(newton_residual_velo * newton_residual_velo + newton_residual_pres * newton_residual_pres);
+
+              // We may need to do a line search if the solution doesn't decrease the norm of the rhs. TODO: more documentation.
+              LinearAlgebra::BlockVector backup_linearization_point = current_linearization_point;
+
+              double test_newton_residual = 0;
+              double test_newton_residual_velo = 0;
+              double test_newton_residual_pres = 0;
+              double lambda = 1;
+              double alfa = 1e-4;
+              unsigned int line_search_iteration = 0;
+
+              do
```

Comment:
Please add a comment here, what this loop does. I have lost track by now in which part of the solver we are.